### PR TITLE
Add mock impl for Adobe Granite Asset API

### PR DESCRIPTION
**Pull request by senn**
Even though it's not the recommended API for assets, the Adobe Granite Asset API has some useful methods compared to the Day CQ Asset API (like removeAsset).  

We use it in some projects (especially the removeAsset functionality).

This PR adds mocks for this.

I opted to use wrappers for the existing mocks (`MockAsset` & `MockAssetManager`) instead of re-implementing the same or similar code. `MockRendition` already had the required methods so that was solved by just adding the other interface.

### CODE REVIEW

**Comment by stefanseifert** [comment]
thanks for the contribution! i applied some cosmetic changes - format code following our standes, add license headers, eliminate some code warnings.

sonarcloud analysis fails due to lack of coverage - can you check https://github.com/wcm-io/io.wcm.testing.aem-mock/pull/49/checks?check_run_id=28941901554 and add a few more coverage methods? thanks!

**Comment by senn** [comment]
Sure, I'll have a look. Thanks for the fast reply.

File: core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapperTest.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+    private static final byte[] BINARY_DATA = new byte[] {
+            0x01, 0x02, 0x03, 0x04, 0x05
+    };
+
+    @Rule
+    public final AemContext context = TestAemContext.newAemContext();
+
+    private Asset asset;
+    private MockAssetManagerTest.DamEventHandler damEventHandler;
```
Comment (henrykuijpers) [inline core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapperTest.java]: It would be nice to move `DamEventHandler` out of `MockAssetManagerTest`, so it can be reused without having to import `MockAssetManagerTest`. It would also be nice to provide the topics when registering the event handler, so that it will already only receive DAM events, the implementation can then be simplified to just expose the list of events (since all other methods are using that same list anyway) - A caller can then use `.stream()` to find the last event or the last 10, ...

File: core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapperTest.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        assertEquals(rendition.getPath(), damEvent.get().getAdditionalInfo());
+
+        asset.removeRendition(renditionName);
+
+        assertNull(asset.getRendition(renditionName));
+        resource = context.resourceResolver().getResource("/content/dam/sample/portraits/scott_reynolds.jpg/jcr:content/renditions/" + renditionName);
+        assertNull(resource);
+
+        damEvent = damEventHandler.getLastEvent();
+        assertTrue(damEvent.isPresent());
```
Comment (henrykuijpers) [inline core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapperTest.java]: Simplifying `DamEventHandler` would result in the `isPresent()` and `get()` calls being removed, which saves some boilerplate

File: core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapperTest.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            if (StringUtils.equals(rendition.getName(), renditionName)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Test
+    public void testRemoveNonExistingRendition() {
+        asset.removeRendition("non-existing");
```
Comment (henrykuijpers) [inline core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapperTest.java]: The result is not tested/asserted

File: core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteRenditionTest.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  public void testProperties() {
+    assertEquals("original", rendition.getName());
+    assertEquals("/content/dam/sample/portraits/scott_reynolds.jpg/jcr:content/renditions/original", rendition.getPath());
+    assertEquals("image/jpeg", rendition.getMimeType());
+    assertNotEquals(0, rendition.hashCode());
+  }
+
+  @Test
+  public void testStream() {
+    assertNotNull(rendition.getStream());
```
Comment (henrykuijpers) [inline core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteRenditionTest.java:59]: I think `Rendition#getStream` is a `@NotNull`-method? It would be nice to verify the actual result value, i.e. by saying which bytes it should be, or that it should be empty (given the `testSize`-method below here)

File: core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteRenditionTest.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  @Test
+  public void testSize() {
+    assertEquals(0L, rendition.getSize());
```
Comment (henrykuijpers) [inline core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteRenditionTest.java:64]: Are we testing here with a 0 byte rendition?

File: core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteRenditionTest.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        .getResource("/content/dam/sample/portraits/scott_reynolds.jpg/jcr:content/renditions/original");
+    this.rendition = resource.adaptTo(Rendition.class);
+  }
+
+  @Test
+  public void testProperties() {
+    assertEquals("original", rendition.getName());
+    assertEquals("/content/dam/sample/portraits/scott_reynolds.jpg/jcr:content/renditions/original", rendition.getPath());
+    assertEquals("image/jpeg", rendition.getMimeType());
+    assertNotEquals(0, rendition.hashCode());
```
Comment (henrykuijpers) [inline core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteRenditionTest.java:54]: This check is not super useful, I would prefer `assertEquals(383948394, rendition.hashCode())`, or is the hashcode not constant? (Because that would be an issue)

**Review by henrykuijpers — CHANGES_REQUESTED** [review]

**Comment by stefanseifert** [comment]
thanks to @henrykuijpers  for the in-depth review! a few cosmetic points are already fixed by my reformatting, most are still relevant.

File: core/src/main/java/io/wcm/testing/mock/aem/dam/MockAsset.java [hunk 7]
```diff
@@ -19,37 +19,20 @@
...
-import java.util.Collection;
-import java.util.Collections;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
-
-import javax.jcr.Binary;
-
+import com.day.cq.commons.jcr.JcrConstants;
+import com.day.cq.dam.api.*;
```
[the remark this answers is not in the document]
Comment (stefanseifert) [reply core/src/main/java/io/wcm/testing/mock/aem/dam/MockAsset.java]: eliminated by reformatting

File: core/src/main/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapper.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+/**
+ * Mock implementation of Adobe Granite {@link Asset}.  This is done by wrapping a {@link MockAsset}
+ */
+@SuppressWarnings("null")
+public class MockGraniteAssetWrapper extends ResourceWrapper implements Asset {
+
+    private final MockAsset asset;
+
+    MockGraniteAssetWrapper(MockAsset asset) {
```
[the remark this answers is not in the document]
Comment (henrykuijpers) [reply core/src/main/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapper.java]: You could use `asset.adaptTo(Resource.class)` to obtain the resource, which should, in the end, obtain the very same Resource?

[the remark this answers is not in the document]
Comment (senn) [reply core/src/main/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapper.java]: fair enough.  a little more verbose but it would indeed be nicer to pass an asset here

File: core/src/main/java/io/wcm/testing/mock/aem/dam/MockAsset.java [hunk 9]
```diff
@@ -80,7 +63,11 @@ class MockAsset extends ResourceWrapper implements Asset {
   @Override
   public <AdapterType> AdapterType adaptTo(Class<AdapterType> type) {
     if (type == Resource.class) {
-      return (AdapterType)resource;
+      return (AdapterType) resource;
+    }
+    //to be able to adapt to granite asset
+    if (type == com.adobe.granite.asset.api.Asset.class) {
```
[the remark this answers is not in the document]
Comment (senn) [reply core/src/main/java/io/wcm/testing/mock/aem/dam/MockAsset.java:86]: makes it a lot harder to access inner resources and assets.  I'd rather not.

File: core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapperTest.java (hunk 2 above)
Comment (senn) [reply core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapperTest.java]: existing code so out of scope for this rather small pr?

File: core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteRenditionTest.java (hunk 6 above)
Comment (senn) [reply core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteRenditionTest.java:54]: the rendition tests are copied from the existing rendition tests (same methods/logic anyway). I did not go into detail about what they actually do.  
This goes for this remark and the other(s) for the MockGraniteRenditionTest class

**Comment by senn** [comment]
I tried to tests for all logic but it's missing some coverage on all the "unsupported oepration" lines ofcourse. My IDE says 78% total coverage, curious to see how much coverage on new lines now...

**Comment by stefanseifert** [comment]
> I tried to add tests for all logic but it's missing some coverage on all the "unsupported oepration" lines ofcourse. My IDE says 78% total coverage, curious to see how much coverage on new lines now...

it still reports 77%. we do not have to test unsupported operations, and also not error handling for exotic edge cases. but i will add a comment or two where i think a coverage is missing or i have an open question.

File: core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteRenditionTest.java (hunk 6 above)
Comment (stefanseifert) [reply core/src/test/java/io/wcm/testing/mock/aem/dam/MockGraniteRenditionTest.java:54]: the original testing code which inspired this is also already very old and not the best one. so, although not ideal, it's fine for me for duplicating it here for the moment. we could improve it here and in the original position in a separate PR.

File: core/src/main/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapper.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+  }
+
+  @Override
+  public String getIdentifier() {
+    return asset.getID();
+  }
+
+  @Override
+  public Rendition setRendition(String s, InputStream inputStream, Map<String, Object> map) {
+    if (map.size() == 1) {
```
Comment (stefanseifert) [inline core/src/main/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapper.java:83]: i do not understand this special logic with map size = 1, and it's also not tested with maps with > 1 entries (and i think using such a map would lead to an unsupported method).

according to https://developer.adobe.com/experience-manager/reference-materials/cloud-service/javadoc/com/adobe/granite/asset/api/Asset.html#setRendition(java.lang.String,java.io.InputStream,java.util.Map) the map is passed through to a RenditionHandler, and that interface defines only two supported properties: https://developer.adobe.com/experience-manager/reference-materials/cloud-service/javadoc/com/adobe/granite/asset/api/RenditionHandler.html

we do not have to mock that handler, but i think for the mock the best solution is to look for an entry with PROPERTY_RENDITION_MIME_TYPE key and ignore the rest of the map. it's not necessarily the first entry.

File: core/src/main/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapper.java [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+      if (val != null) {
+        return (Rendition)asset.addRendition(s, inputStream, val.toString());
+      }
+    }
+    return (Rendition)asset.addRendition(s, inputStream, map);
+  }
+
+  @Override
+  public Rendition setRendition(String s, Binary binary, Map<String, Object> map) {
+    return (Rendition)asset.addRendition(s, binary, map);
```
Comment (stefanseifert) [inline core/src/main/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapper.java:94]: this routes to an unsupported method, so the missing coverage is misleading. maybe more transparent if we just move this method to unsupported as well.

File: core/src/main/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapper.java (hunk 10 above)
Comment (stefanseifert) [reply core/src/main/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapper.java:83]: i will quickly apply the changes here and update the test cases

Comment (senn) [reply core/src/main/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapper.java]: It was a way of being able to pass only thr mime type but your solution is more elegant.

File: core/src/main/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapper.java (hunk 11 above)
Comment (senn) [reply core/src/main/java/io/wcm/testing/mock/aem/dam/MockGraniteAssetWrapper.java]: The idea was that when the delegate method would eventually be implemented, so would this.
